src/app/ngrx/reducers/podcast-metrics.reducer.ts

```diff
@@ -0,0 +1,62 @@
+import { Action } from '@ngrx/store';
+import ActionTypes from '../actions/action.types';
+import { PodcastMetricsModel } from '../model';
+import { subtractDataset } from './metrics.util';
+
+const initialState = [];
+
+export function PodcastMetricsReducer(state: PodcastMetricsModel[] = initialState, action: Action) {
+  let podcastIdx: number, podcast: PodcastMetricsModel, newState: PodcastMetricsModel[], metricsProperty: string;
+  switch (action.type) {
+    case ActionTypes.CASTLE_PODCAST_METRICS:
+      metricsProperty = action.payload.filter.interval.key
+        + action.payload.metricsType.charAt(0).toUpperCase()
+        + action.payload.metricsType.slice(1);
+      podcastIdx = state.findIndex(p => p.seriesId === action.payload.podcast.seriesId);
+      if (podcastIdx > -1) {
+        podcast = Object.assign({}, state[podcastIdx]);
+        podcast[metricsProperty] = action.payload.metrics;
+        podcast[metricsProperty + 'Others'] = [...podcast[metricsProperty]];
+        podcast.episodeIdsNotInOthers = [];
+        newState = [...state.slice(0, podcastIdx), podcast, ...state.slice(podcastIdx + 1)];
+      } else {
+        podcast = {
+          seriesId: action.payload.podcast.seriesId,
+          feederId: action.payload.podcast.feederId
+        };
+        podcast[metricsProperty] = action.payload.metrics;
+        podcast[metricsProperty + 'Others'] = [...podcast[metricsProperty]];
+        podcast.episodeIdsNotInOthers = [];
+        newState = [podcast, ...state];
+      }
+      // console.log('PodcastMetricsReducer', action.type, newState);
+      return newState;
+    case ActionTypes.CASTLE_EPISODE_METRICS:
+      metricsProperty = action.payload.filter.interval.key
+        + action.payload.metricsType.charAt(0).toUpperCase()
+        + action.payload.metricsType.slice(1);
+      podcastIdx = state.findIndex(p => p.seriesId === action.payload.episode.seriesId);
+      if (podcastIdx > -1 &&
```

Comment:
This has only got to handle all episodes that are in PodcastMetrics state, present in the filter, and not already subtracted from Others. I haven't yet worked out what it will need to do when I add episode selection, will probably have to reset and recalculate but I'm just not that far yet.

Reply:
I don't know if I described that well. This store current has two sets of metrics, the podcast's dailyDownloads and dailyDownloadsOthers. It's only responsibility for episode metrics is to subtract them from the podcast dailyDownloads into dailyDownloadsOthers and keep track of the fact that they have been subtracted. The episode metrics store is responsible for holding those episode metrics.

Reply:
👍